clang/lib/3C/RewriteUtils.cpp

```diff
@@ -229,11 +229,52 @@ static void emit(Rewriter &R, ASTContext &C) {
       };
 
       // Check whether we are allowed to write this file.
+      //
+      // In our testing as of 2021-03-15, the file path returned by
+      // FE->tryGetRealPathName() was canonical, but be safe against unusual
+      // situations or possible future changes to Clang before we actually write
+      // a file. We can't use FE->getName() because it seems it may be relative
+      // to the `directory` field of the compilation database, which (now that
+      // we no longer use `ClangTool::run`) is not guaranteed to match 3C's
+      // working directory.
       std::string ToConv = FE->tryGetRealPathName().str();
       std::string FeAbsS = "";
-      getCanonicalFilePath(ToConv, FeAbsS);
+      std::error_code EC = tryGetCanonicalFilePath(ToConv, FeAbsS);
+      if (EC) {
+        unsigned ErrorId = DE.getCustomDiagID(
+            UnwritableChangeDiagnosticLevel,
+            "3C internal error: not writing the new version of this file due "
+            "to failure to re-canonicalize the file path provided by Clang");
+        DE.Report(SM.translateFileLineCol(FE, 1, 1), ErrorId);
+        {
+          unsigned NoteId =
+              DE.getCustomDiagID(DiagnosticsEngine::Note,
+                                 "file path from Clang was %0; error was: %1");
+          auto ErrorBuilder = DE.Report(NoteId);
+          ErrorBuilder.AddString(ToConv);
+          ErrorBuilder.AddString(EC.message());
+        }
+        PrintExtraUnwritableChangeInfo();
+        continue;
+      }
+      if (FeAbsS != ToConv) {
```

Comment:
Was the purpose of `tryGetRealPathName()` above to convert or to check that no conversion is required? The rest of the code only uses `FeAbsS` so it seems like this diagnostic is overkill. Maybe leave the note and remove the error?

Reply:
We expect `tryGetRealPathName` to return a canonical path, meaning that `tryGetCanonicalFilePath` would not change it.  If the path does change, then the use of the old path elsewhere in the program analysis (e.g., to determine which elements need to be constrained because they are outside the base directory) may have caused incorrect behavior, so it would be good to report an error.  It would probably be safe to write the output file anyway if you prefer that; the error (and nonzero exit code) is the important thing.

Reply:
Ok. I'd prefer to have this type of error at some other point in the codebase, though. Like if we can't canonicalize, just exit after the initial clang pass (or before). But I'll remove the hold if you think this is a good place for it to be.

Reply:
I agree, this is a little awkward, but I don't know of a better place where we have access to a list of files that is guaranteed to include any file that 3C might try to modify.  It looks like we might have to either get a list of all files in the `SourceManager` or somehow hook the process of adding rewrite buffers to the `Rewriter`.

If the problem is just that this code is getting so long that it makes it hard to follow the high-level flow of `emit`, I can move it to a subroutine.

Reply:
It's up to you now, I ran out of suggestions ;)

Reply:
I'll leave it for now then.